File: app/src/main/java/org/dhis2/usescases/reservedValue/ReservedValueModel.java
package org.dhis2.usescases.reservedValue;

import androidx.annotation.NonNull;

import com.google.auto.value.AutoValue;

import javax.annotation.Nullable;

@AutoValue
public abstract class ReservedValueModel {

    @NonNull
    public abstract String uid();

    @Nullable
    public abstract String orgUnitUid();

    @NonNull
    public abstract Boolean patternContainsOU();

    @Nullable
    public abstract String orgUnitName();

    @NonNull
    public abstract String displayName();

    @NonNull
    public abstract Integer reservedValues();

    @NonNull
    public static ReservedValueModel create(@NonNull String uid,
                                            @NonNull String displayName,
                                            boolean patternContainsOU,
                                            @Nullable String orgUnitUid, @Nullable String orgUnitName,
                                            int reservedValue) {
        return new AutoValue_ReservedValueModel(uid, orgUnitUid, patternContainsOU, orgUnitName, displayName, reservedValue);
    }

}

File: app/src/main/java/org/dhis2/usescases/reservedValue/ReservedValueModel.kt
package org.dhis2.usescases.reservedValue

import io.reactivex.processors.FlowableProcessor

data class ReservedValueModel(
    val attributeUid: String,
    val attributeName: String,
    val orgUnitUid: String?,
    val orgUnitName: String?,
    val count: Int,
    val leftValuesLabel: String,
    val refillProcessor: FlowableProcessor<String>,
) {
    fun hasOrgUnit(): Boolean {
        return orgUnitUid != null
    }

    fun valuesLeft(): String {
        return leftValuesLabel.format(count)
    }

    fun refill() {
        refillProcessor.onNext(attributeUid)
    }
}
